pkg/pool-gyro/contracts/GyroECLPPool.sol

```diff
@@ -0,0 +1,229 @@
+        Vector2 memory invariant;
+        {
+            (int256 currentInvariant, int256 invErr) = GyroECLPMath.calculateInvariantWithError(
+                request.balancesScaled18,
+                eclpParams,
+                derivedECLPParams
+            );
+            // invariant = overestimate in x-component, underestimate in y-component
+            // No overflow in `+` due to constraints to the different values enforced in GyroECLPMath.
+            invariant = Vector2(currentInvariant + 2 * invErr, currentInvariant);
+        }
+
+        if (request.kind == SwapKind.EXACT_IN) {
+            uint256 amountOutScaled18 = GyroECLPMath.calcOutGivenIn(
+                request.balancesScaled18,
+                request.amountGivenScaled18,
+                tokenInIsToken0,
+                eclpParams,
+                derivedECLPParams,
+                invariant
+            );
+
+            return amountOutScaled18;
+        } else {
+            uint256 amountInScaled18 = GyroECLPMath.calcInGivenOut(
+                request.balancesScaled18,
+                request.amountGivenScaled18,
+                tokenInIsToken0,
+                eclpParams,
+                derivedECLPParams,
+                invariant
+            );
+
+            return amountInScaled18;
+        }
+    }
+
+    /** @dev reconstructs ECLP params structs from immutable arrays */
+    function _reconstructECLPParams() private view returns (EclpParams memory params, DerivedEclpParams memory d) {
+        (params.alpha, params.beta, params.c, params.s, params.lambda) = (
+            _paramsAlpha,
+            _paramsBeta,
+            _paramsC,
+            _paramsS,
+            _paramsLambda
+        );
+        (d.tauAlpha.x, d.tauAlpha.y, d.tauBeta.x, d.tauBeta.y) = (_tauAlphaX, _tauAlphaY, _tauBetaX, _tauBetaY);
+        (d.u, d.v, d.w, d.z, d.dSq) = (_u, _v, _w, _z, _dSq);
+    }
+
+    function getECLPParams() external view returns (EclpParams memory params, DerivedEclpParams memory d) {
+        return _reconstructECLPParams();
+    }
+
+    /// @inheritdoc ISwapFeePercentageBounds
+    function getMinimumSwapFeePercentage() external pure returns (uint256) {
+        // Liquidity Approximation tests shows that add/remove liquidity combinations are more profitable than a swap
+        // if the swap fee percentage is 0%, which is not desirable. So, a minimum percentage must be enforced.
+        return 1e12; // 0.000001%
+    }
+
+    /// @inheritdoc ISwapFeePercentageBounds
+    function getMaximumSwapFeePercentage() external pure returns (uint256) {
+        return 1e18; // 100%
+    }
+
+    /// @inheritdoc IUnbalancedLiquidityInvariantRatioBounds
+    function getMinimumInvariantRatio() external pure returns (uint256) {
+        return 0;
+    }
+
+    /// @inheritdoc IUnbalancedLiquidityInvariantRatioBounds
+    function getMaximumInvariantRatio() external pure returns (uint256) {
+        return type(uint256).max;
+    }
```

Comment:
Let's double check these 2 as well. Setting reasonable boundaries, even if not needed, is never a bad idea.